src/types.rs

```diff
@@ -33,21 +33,101 @@ impl TypeRef {
+#[derive(Debug, Clone, PartialEq)]
+pub enum TypeDefinition {
+    NamedTypeDefinition(NamedTypeDefinition),
+    AnonymousTypeDefinition(AnonymousTypeDefinition),
+}
+
+/// A [AnonymousTypeDefinition] consists of a name and zero or more constraints.
 #[derive(Debug, Clone)]
-pub struct TypeDefinition {
-    name: Option<String>,
+pub struct AnonymousTypeDefinition {
     constraints: Vec<Constraint>,
 }
 
-impl TypeDefinition {
-    pub fn new(name: Option<String>, constraints: Vec<Constraint>) -> Self {
+impl AnonymousTypeDefinition {
+    pub fn new(constraints: Vec<Constraint>) -> Self {
+        Self { constraints }
+    }
+
+    pub fn constraints(&self) -> &[Constraint] {
+        &self.constraints
+    }
+
+    /// Parse constraints inside an [OwnedStruct] to a schema [Type]
+    pub fn parse_from_isl_type_and_update_type_store(
+        isl_type: &AnonymousIslType,
+        type_store: &mut TypeStore,
+        pending_types: &mut PendingTypes,
+    ) -> IonSchemaResult<Self> {
+        let mut constraints = vec![];
+
+        // add this unresolved type to context for type_id
+        let type_id = pending_types.add_type(type_store);
+
+        // convert IslConstraint to Constraint
+        for isl_constraint in isl_type.constraints() {
+            let constraint = match isl_constraint {
+                IslConstraint::AllOf(type_references) => {
+                    let all_of: AllOfConstraint = AllOfConstraint::resolve_from_isl_constraint(
+                        type_references,
+                        type_store,
+                        pending_types,
+                    )?;
+                    Constraint::AllOf(all_of)
+                }
+                IslConstraint::Type(type_reference) => {
+                    let type_constraint: TypeConstraint =
+                        TypeConstraint::resolve_from_isl_constraint(
+                            type_reference,
+                            type_store,
+                            pending_types,
+                        )?;
+                    Constraint::Type(type_constraint)
+                }
+            };
+            constraints.push(constraint);
+        }
+
+        let type_def = AnonymousTypeDefinition::new(constraints);
+
+        // update with this resolved type_def to context for type_id
+        pending_types.update_anonymous_type(type_id, type_def.to_owned(), type_store);
+
+        Ok(type_def)
+    }
+}
+
+impl PartialEq for AnonymousTypeDefinition {
+    fn eq(&self, other: &Self) -> bool {
+        self.constraints == other.constraints()
+    }
+}
+
+impl TypeValidator for AnonymousTypeDefinition {
+    fn is_valid(&self, value: &OwnedElement) -> bool {
+        todo!()
+    }
+
+    fn validate(&self, value: &OwnedElement, issues: &mut Violations) {
+        todo!()
+    }
+}
+
+/// A [NamedTypeDefinition] consists of a name and zero or more constraints.
+#[derive(Debug, Clone)]
+pub struct NamedTypeDefinition {
```

Comment:
`NamedTypeDefinition` and `AnonymousTypeDefinition` are so closely related that having two separate structs with nearly identical method implementations feels like overkill to me. Here are some alternatives to consider.

--------

### Option 1

We could restructure the `TypeDefinition` enum to hold the name:

```rust
#[derive(Debug, Clone, PartialEq)]
pub enum TypeDefinition {
    Named(String, TypeDefinition), // This variant has a `String` for the name
    Anonymous(TypeDefinition),
}
```

With this option, we can still distinguish between the named and anonymous cases via the enum but we eliminate a largely redundant type.

--------

### Option 2

This requires a bit more effort but would still simplify things quite a bit; we could redefine `NamedTypeDefinition` as:
```rust
pub struct NamedTypeDefinition {
    name: String,
    type_definition: TypeDefinition
}
```
We still have a second full type (`NamedTypeDefinition`), but by re-using `TypeDefinition` via composition, we allow all of the methods/functionality to go on living on the `TypeDefinition` type itself.

--------

My personal preference is option 1.